### PR TITLE
[RDY]"Look at" with long descriptions (of critters, furniture etc.)

### PR DESCRIPTION
**Pull request by Coolthulhu**
Multiple times I wanted to describe something in the game, but couldn't find a good menu for it.
Terrain, furniture, critters and NPCs have this tiny window on the right, where only a single sentence of compact prose will fit.

My idea is as follows:
* When looking at a tile, you can press a button to open an extended description
* This opens a tabbed, full-screen window that allows checking different objects/features on the tile
* Each supported feature then generates some description that could take a lot of space, contain trivia, very specific details etc.

For example:
* You see an apple tree here. In autumn, you could gather 4-6 apples from it. "Apples of New England are known for their x-es and y-s."
* You see a zombie here. It is moderately damaged. It has been stripped of gear. It is moving slower than usual. It is looking at something to your left but you can't see what. "A reanimated corpse that wants to kill you." "Thanks to the notes in lab, you know that those creatures are animated by ooze from outer space." It can bash down obstacles. It can bite. It doesn't need to breathe.
* You see Urist McSoapmaker here. He is neutral and wants to talk. He is wearing socks, socks (fits), and a trenchcoat. He is unarmed. He has 8-14 strength and 6-16 dexterity. He has the following traits: fleet-footed, albino. He may have more traits that you can't see. He looks completely wasted. He is looking at you. His right arm is `||\..` and infected.

Basically, a place to dump all the extra info that would be amazing to display, but that doesn't fit anywhere due to how is the UI structured.

___

In this PR, I added:
* The description menu
* HP and equipment display for NPCs
* Just name and HP for monsters
* Just the name for furniture/terrain

### CODE REVIEW

**Comment by Zireael07** [comment]
Most roguelikes have some sort of such descriptions in a separate window. +100 from me.

**Comment by BrettDong** [comment]
How those detailed descriptions are generated?

For the zombies:
>You see a zombie here. It is moderately damaged. It has been stripped of gear. It is moving slower than usual. It is looking at something to your left but you can't see what. "A reanimated corpse that wants to kill you." "Thanks to the notes in lab, you know that those creatures are animated by ooze from outer space." It can bash down obstacles. It can bite. It doesn't need to breathe.

Is that you prepare a number of sentences for every type of depiction ("It need to breathe", "It doesn't need to breathe") and select which to appear according to the situation then combine those sentences to form the entire paragraph?

If you generate those sentences in the code like `_( "It %s to breathe" ), need_breathe ? "need" : "doesn't need"`, it might be difficult or even impossible to localize. So I recommend preparing a bunch of fixed strings and select which to display instead of generating text following the English grammar.

**Comment by Coolthulhu** [comment]
> How those detailed descriptions are generated?

I'd expect some specific monster flags, abilities and states to have descriptions paired to them.

For example, no breathing flag would have a "It doesn't need to breathe", while the default (it does need to breathe) would likely be omitted.

> it might be difficult or even impossible to localize

In those cases we generally expose the whole sentence to the translator, without chopping it up into pieces. That is, translator would see "It does need to breathe" or "It doesn't need to breathe".

**Comment by keyspace** [comment]
Trying out to-merge branch.

Fails to compile for me with

```
src/character.cpp: In member function ‘virtual std::__cxx11::string Character::extended_description() const’:
src/character.cpp:2252:22: error: ‘accumulate’ is not a member of ‘std’
     size_t longest = std::accumulate( bps.begin(), bps.end(), 0, []( size_t m, body_part bp ) {
                      ^~~
make: *** [Makefile:685: obj/character.o] Error 1
```

without

``` c++
#include <numeric>
```

added to `character.cpp`.

Using `gcc (GCC) 6.3.1 20170109` on Arch Linux 64-bit.

File: src/character.cpp [hunk 1]
```diff
@@ -2142,3 +2143,151 @@ long Character::ammo_count_for( const item &gun )
...
+        case hp_leg_r:
+            return bp_leg_r;
+        default:
+            return num_bp;
+    }
+}
+
+body_part Character::get_random_body_part( bool main ) const
+{
+    // TODO: Refuse broken limbs, adjust for mutations
```
Comment (keyspace) [inline src/character.cpp]: Nit: PR no longer [WiP], change to `@todo`?..

Comment (Coolthulhu) [inline src/character.cpp]: That's an incredibly long-term todo.

File: src/descriptions.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                    const ter_id tid = m.ter( p );
+                    desc = tid.obj().description();
+                }
+                break;
+        }
+
+        werase( w_main );
+        fold_and_print_from( w_main, 0, 0, width, 0, c_ltgray, desc );
+        wrefresh( w_main );
+        // TODO: use input context
```
Comment (keyspace) [inline src/descriptions.cpp]: Non-issue: doing it like this actually allows the key bindings to work, even in the face of recent input-related trouble!

File: src/monster.cpp [hunk 3]
```diff
@@ -385,6 +385,33 @@ void monster::get_Attitude(nc_color &color, std::string &text) const
     }
 }
 
+std::pair<std::string, nc_color> hp_description( int cur_hp, int max_hp )
+{
+    std::string damage_info;
+    nc_color col;
+    if( cur_hp >= max_hp ) {
+        damage_info = _("It is uninjured");
```
Comment (keyspace) [inline src/monster.cpp]: Nit: no dots at end of sentence looks OK in non-extended view, but not so well in extended.

**Comment by keyspace** [comment]
Exiting the extended view menu via ESC does not redraw most of the right-side menu. Only the tiny description window is present, not the mini-map, message log, etc.. Not a big issue, but very unusual.

This on ncurses.

File: src/monster.cpp [hunk 4]
```diff
@@ -439,6 +447,16 @@ int monster::print_info(WINDOW* w, int vStart, int vLines, int column) const
     return vStart;
 }
 
+std::string monster::extended_description() const
+{
+    // @todo Add tons of info here
+    std::string ret = string_format( _ ( "This is a %s" ), name().c_str() );
```
Comment (BrettDong) [inline src/monster.cpp]: A redundant space is inserted: `_ ( "This is...`

**Comment by BrettDong** [comment]
Suggestions on the UI design:
* This could be a centered window like inventory. Space might be insufficient to show the entire text, so a vertical scrollbar might be needed.
* [Creature], [Terrain] and [Furniture] could be three tabs similar to mission window.

**Comment by Coolthulhu** [comment]
* I don't like those centered windows and would prefer to avoid them whenever possible.
* Actually displayed tabs could be useful, I just didn't want a huge UI system for the first PR.

It would be pretty useful to have a tabbed UI class. At the moment all tabbed UIs are reimplementing the wheel.

**Comment by keyspace** [comment]
> I just didn't want a huge UI system for the first PR.

The `e` binding isn't even displayed when viewing, and there's little useful info ATM, so can be considered a WIP feature.

> It would be pretty useful to have a tabbed UI class.

That could solve it for people who like either full-screen or not. Massive TODO, of course.

File: src/game.cpp [hunk 5]
```diff
@@ -8684,6 +8685,10 @@ tripoint game::look_around( WINDOW *w_info, const tripoint &start_point,
             if( !MAP_SHARING::isCompetitive() || MAP_SHARING::isDebugger() ) {
                 display_scent();
             }
+        } else if( action == "EXTENDED_DESCRIPTION" ){
```
Comment (BrettDong) [inline src/game.cpp]: Minor: a space missing

File: src/descriptions.cpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        switch( cur_target ) {
+            case description_target::creature: {
+                const Creature *critter = critter_at( p );
+                if( critter != nullptr && u.sees( *critter ) ) {
+                    desc = critter->extended_description();
+                } else {
+                    desc = _( "You do not see any creature here." );
+                }
+            }
+            break;
```
Comment (BrettDong) [inline src/descriptions.cpp]: Minor: inconsistent `break;` statement indention

Comment (BrettDong) [inline src/descriptions.cpp]: Minor: inconsistent `break;` statement indention

Comment (Coolthulhu) [inline src/descriptions.cpp]: The file is astyled, the breaks are as astyle wants them.

Comment (BrettDong) [inline src/descriptions.cpp]: Oh then that's fine. I only found those code style related minor issues, no other obvious problem exists in the code as far as I read. BTW I don't know why my comment got posted twice.

**Comment by kevingranade** [comment]
Not asking for a change in this PR, but if the extended description is going to be full window, it might as well show all of the elements at once.  Right now it's a gigantic expanse of blank screen.

**Comment by Night-Pryanik** [comment]
How the extended menu is activated? I press `e`, but the game asks the usual "Examine where?".

**Comment by BrettDong** [comment]
In `x` look around mode, move the cursor to a tile and then press `e`.

**Comment by Night-Pryanik** [comment]
@BrettDong thanks.
It's hard to find, though, without any hints on it.

**Comment by Coolthulhu** [comment]
Where should I add hints?

**Comment by Night-Pryanik** [comment]
> Where should I add hints?


Maybe another line here? Smth like "Press e to get extended info".
